sdk/core/System.ClientModel/src/Pipeline/MessageLoggingPolicy.cs

```diff
@@ -0,0 +1,342 @@
+                {
+                    responseBytes = new byte[_maxLength];
+                    response.ContentStream.Read(responseBytes, 0, _maxLength);
+                    response.ContentStream.Seek(0, SeekOrigin.Begin);
+                }
+
+                if (response.IsError)
+                {
+                    _messageLogger.LogErrorResponseContent(requestId, responseBytes, responseTextEncoding);
+                }
+                else
+                {
+                    _messageLogger.LogResponseContent(requestId, responseBytes, responseTextEncoding);
+                }
+            }
+            else
+            {
+                response.ContentStream = new LoggingStream(_messageLogger, requestId, _maxLength, response.ContentStream, response.IsError, responseTextEncoding);
+            }
+        }
+    }
+
+    #region MaxLengthStream
+    private class MaxLengthStream : MemoryStream
+    {
+        private int _bytesLeft;
+
+        public MaxLengthStream(int maxLength) : base()
+        {
+            _bytesLeft = maxLength;
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            DecrementLength(ref count);
+            if (count > 0)
+            {
+                base.Write(buffer, offset, count);
+            }
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return count > 0 ? base.WriteAsync(buffer, offset, count, cancellationToken) : Task.CompletedTask;
+        }
+
+        private void DecrementLength(ref int count)
+        {
+            var left = Math.Min(count, _bytesLeft);
+            count = left;
+
+            _bytesLeft -= count;
+        }
+    }
+    #endregion
+
+    #region LoggingStream
+    private class LoggingStream : Stream
+    {
+        private readonly string _requestId;
+        private int _maxLoggedBytes;
+        private readonly int _originalMaxLength;
+        private readonly Stream _originalStream;
+        private readonly bool _error;
+        private readonly Encoding? _textEncoding;
+        private int _blockNumber;
+        private readonly PipelineMessageLogger _messageLogger;
+
+        public LoggingStream(PipelineMessageLogger messageLogger, string requestId, int maxLoggedBytes, Stream originalStream, bool error, Encoding? textEncoding)
+        {
+            // Should only wrap non-seekable streams
+            Debug.Assert(!originalStream.CanSeek);
+            _requestId = requestId;
+            _maxLoggedBytes = maxLoggedBytes;
+            _originalMaxLength = maxLoggedBytes;
+            _originalStream = originalStream;
+            _error = error;
+            _textEncoding = textEncoding;
+            _messageLogger = messageLogger;
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            return _originalStream.Seek(offset, origin);
```

Comment:
Why delegate here if you're enforcing a non-seekable stream?    Conversely, why enforce a non-seekable stream if you're going to delegate here?

We throw `NotSupportedException` explicitly for write operations rather than delegating.

Reply:
I'll update (this was also a copy/paste from Core). I think since CanSeek is enforced as false presumably the original stream will throw NotSupportedException so idk if it is clearer to have the original stream do that.